### DISCUSSION TITLE
Google Cloud Translation API v2 support

### DISCUSSION THREAD

**Discussion by Alexamenus** [ogkalu2/comic-translate#180] [Ideas]
Right now all that is supported is v1. v2 is also free and has much higher translation quality similar to GPT. Could support be added in the future?

**Comment by BeeWith** [comment]
This would be great! I've noticed the translations that come from the API now are often bad for Japanese. But if I drop the Japanese into Google's Translate website, I get much better results.

**Comment by orbitdesign** [comment]
can the option of using Google Translate v2 be added into the app? 
Google Translate v1 is really bad in translation.

**Comment by BeeWith** [comment]
Looks like Ogkalu added a spot for the Gemini API key at some point. I'm testing now and it seems to return the same results as Google Translate for text entry. That's a nice improvement and still free. 

I have noticed that the Google Translate Image website still shows different results. I suspect this is partly due to not removing returns (so you get separate chunks of text, which is sometimes better, sometimes worse) and due to better OCR (there are some comics Ogkalu Comic Translate still can't handle, but Google Image Translate can).

In any case, if you'd like to try Gemini, login to your Google account, go to https://aistudio.google.com/, accept the terms and conditions, accept the default project, click the "Create API Key" button. Copy the key and paste it into Ogkalu Comic Translate (click the Settings Cog, click the Credentials tab, paste in the Gemini field). 

That's all I needed to do, but I did previously have the original API setup, so if you have trouble, you may need to do that.

**Comment by BeeWith** [comment]
Update. Results with Gemini seemed better on the first comic I tried, but following that I found some that pretty much returned results that were the same as Ogkalu. I had trouble finding answers to why that might be. Finally I failed out and tried asking Gemini from the AIStudio page. For reference, I've never gotten a correct answer from an AI, so I'm not sure how much I trust this one, but I found it interesting. 

The short version of why API results are different from Google Translate Website results is that they are doing more processing with the web results, and often using beta versions of more advanced models (so it's live testing). The API uses stable models and does less processing (for example, no spell checking). It's intended for massive data dumps, so it's also optimized to use less power and more speed when translating. It also doesn't use the glossaries that it's generated from the web translations, so the API gives raw machine translations (unless you setup your own glossary).

Recommendations for improving API results were:

1) Switch to the Advanced API (v3): This gives you access to the newest models. (This probably costs money, so won't work for my purposes.)

2) Use the Translation LLM: Use the model ID general/translation-llm. It is powered by Gemini and is specifically tuned for higher-quality, context-aware translations. (This is interesting, because I thought I was using Gemini in Ogkalu... but perhaps not.)

3) Implement a Glossary: If there are specific industry terms the API is getting wrong, the Advanced API allows you to upload a glossary to force specific translations.

4) Pre-clean your data: Use a separate spell-checker or LLM to clean up typos and slang before sending the text to the Translation API.
